File: Week_01/977.有序数组的平方.c
/*
 * @lc app=leetcode.cn id=977 lang=c
 *
 * [977] 有序数组的平方
 */

// @lc code=start


/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* sortedSquares(int* A, int ASize, int* returnSize){
    if (A == NULL || ASize == 0)
    {
        *returnSize = 0;
        return NULL;
    }
    int left = 0;
    int right = ASize - 1;
    int i;
    int *res = (int *)malloc(sizeof(int) * ASize);
    for (i = ASize - 1; i >= 0; i--)
    {
        if(abs(A[left]) > abs(A[right]))
        {
            res[i] = A[left] * A[left];
            left++;
        }
        else
        {
            res[i] = A[right] * A[right];
            right--;
        }
    }
    *returnSize = ASize;
    return res;
}


// @lc code=end

File: Week_01/977.有序数组的平方.java
/*
 * @lc app=leetcode.cn id=977 lang=java
 *
 * [977] 有序数组的平方
 */

// @lc code=start
class Solution {
    public int[] sortedSquares(int[] A) {
        if (A == null || A.length == 0) {
            return new int[0];
        }

        int len = A.length;
        int left = 0;
        int right = len - 1;
        int[] res = new int[len];
        for (int i = len - 1; i >= 0; i--) {
            if (Math.abs(A[left]) > Math.abs(A[right])) {
                res[i] = A[left] * A[left];
                left++;
            } else {
                res[i] = A[right] * A[right];
                right--;
            }
        }
        return res;
    }
}
// @lc code=end

